extensions-contrib/druid-jwt-extensions/src/main/java/org/apache/druid/security/jwt/JwtJwksAuthenticator.java

```diff
@@ -0,0 +1,207 @@
+import org.jose4j.jwt.MalformedClaimException;
+import org.jose4j.jwt.consumer.InvalidJwtException;
+
+import javax.annotation.Nullable;
+import javax.servlet.DispatcherType;
+import javax.servlet.Filter;
+import javax.servlet.FilterChain;
+import javax.servlet.FilterConfig;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.util.EnumSet;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+
+@JsonTypeName("jwt_jwks")
+public class JwtJwksAuthenticator implements Authenticator
+{
+  private static final Logger LOG = new Logger(JwtJwksAuthenticator.class);
+
+  private final String name;
+  private final String authorizerName;
+  private final String issuer;
+  private final Set<String> audience;
+  private final String jwksUri;
+
+  private final JwtVerifier jwtVerifier;
+  private final IssuerExtractor issuerExtractor;
+
+  @JsonCreator
+  public JwtJwksAuthenticator(
+      @JsonProperty("name") String name,
+      @JsonProperty("issuer") String issuer,
+      @JsonProperty("audience") Set<String> audience,
+      @JsonProperty("jwksUri") String jwksUri,
+      @JsonProperty("authorizerName") String authorizerName
+  )
+  {
+    this.name = Preconditions.checkNotNull(name, "null Name");
+    this.issuer = Preconditions.checkNotNull(issuer, "null issuer");
+    this.jwksUri = Preconditions.checkNotNull(jwksUri, "null jwksUri");
+    this.authorizerName = authorizerName;
+    this.audience = audience;
+    this.jwtVerifier = new JwtVerifier(issuer, jwksUri, audience);
+    this.issuerExtractor = new IssuerExtractor();
+  }
+
+  @Override
+  public Filter getFilter()
+  {
+    return new HTTPAuthenticationFilter();
+  }
+
+  @Override
+  public Class<? extends Filter> getFilterClass()
+  {
+    return HTTPAuthenticationFilter.class;
+  }
+
+  @Override
+  public String getAuthChallengeHeader()
+  {
+    return "Bearer";
+  }
+
+  @Override
+  public Map<String, String> getInitParameters()
+  {
+    return null;
+  }
+
+  @Override
+  public String getPath()
+  {
+    return "/*";
+  }
+
+  @Override
+  public EnumSet<DispatcherType> getDispatcherType()
+  {
+    return null;
+  }
+
+
+  @Override
+  @Nullable
+  public AuthenticationResult authenticateJDBCContext(Map<String, Object> context)
+  {
+    String bearerToken = (String) context.get("bearer-token");
+
+    if (bearerToken != null) {
+      try {
+        JwtClaims jwtClaims = jwtVerifier.authenticate(bearerToken);
+        if (jwtClaims != null) {
+          LOG.info("JWT validation succeeded!");
+          return new AuthenticationResult(
+              jwtClaims.getSubject(),
+              authorizerName,
+              name,
+              null
+          );
+        }
+      }
+      catch (InvalidJwtException | MalformedClaimException ex) {
```

Comment:
it might be worth separating the two cases here to distinguish between a bad token and one that's missing the necessary claims